src/hl/plist/dataset_create.rs

```diff
@@ -741,11 +741,13 @@ impl DatasetCreate {
     pub fn get_fill_value_as<T: H5Type>(&self) -> Result<Option<T>> {
         let dtype = Datatype::from_type::<T>()?;
         Ok(self.get_fill_value(&dtype.to_descriptor()?)?.map(|value| unsafe {
-            let mut out: T = mem::zeroed();
+            let mut out: mem::MaybeUninit<T> = mem::MaybeUninit::uninit();
             let buf = value.get_buf();
-            ptr::copy_nonoverlapping(buf.as_ptr(), (&mut out as *mut T).cast(), buf.len());
-            mem::forget(value);
-            out
+            ptr::copy_nonoverlapping(buf.as_ptr(), out.as_mut_ptr().cast(), buf.len());
+            // Drop the Box<[u8]>, but not subfields
+            value.drop_nonrecursive();
+            // We now have exclusive access to all subfields
+            out.assume_init()
```

Comment:
Since `drop_nonrecursive()` is only used here, I wonder if it could be made `pub(crate)` and this whole block could be pulled out into a method on `OwnedDynValue` like this:

```rust
    #[doc(hidden)]
   pub unsafe fn cast<T: H5Type>(mut self) -> Result<T> {
        ...
   }
```

where it could also check that `T::type_descriptor() == self.tp` (or it could return `-> T` without `Result` and just panic on mismatch).

And then here it just becomes

```rust
let dtype = Datatype::from_type::<T>()?;
self.get_fill_value(&dtype.to_descriptor()?)?.map(|v| unsafe { v.cast() }).transpose()
```

Reply:
The new code pushes all the unsafety on `hdf5-types`, making the code a lot clearer. As zero-sized fixed arrays are not possible to create in `hdf5`, I've added a small expection to `dyn_drop` of `DynArray` to support this code path.